apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/concurrent/ConcurrentRdaSink.java

```diff
@@ -153,16 +277,39 @@ public int writeClaims(Collection<TClaim> objects) throws ProcessingException {
 
   /** {@inheritDoc} */
   @Override
-  public int getProcessedCount() throws ProcessingException {
-    return writerPool.getProcessedCount();
+  public int getProcessedCount() {
+    return unreportedProcessedCount.getAndSet(0);
   }
 
   /** Shuts down the thread pool. Any unwritten data is flushed to the database. */
   @Override
   public void shutdown(Duration waitTime) throws ProcessingException {
+    log.info("shutdown called");
+    if (running.getAndSet(false)) {
+      publisher.complete();
+      log.info("shutdown emit complete");
+    }
     try {
-      writerPool.shutdown(waitTime);
+      final var closer = new MultiCloser();
+      log.info("shutdown wait for latch");
+      closer.close(() -> waitForLatch(waitTime));
+      for (ClaimWriter<TMessage, TClaim> claimWriter : claimWriters) {
+        log.info("shutdown close claimWriter {}", claimWriter.getId());
+        closer.close(claimWriter::close);
+      }
+      log.info("shutdown close sequenceWriter");
+      closer.close(sequenceNumberWriter::close);
+      log.info("shutdown close sink");
+      closer.close(sink::close);
+      log.info("shutdown check for errors");
+      closer.close(this::throwIfErrorPresent);
+      log.info("shutdown finish");
+      closer.finish();
+    } catch (ProcessingException ex) {
+      log.error("shutdown failed: ex={}", ex.getMessage(), ex);
+      throw ex;
     } catch (Exception ex) {
+      log.error("shutdown failed: ex={}", ex.getMessage(), ex);
```

Comment:
lots of steps for the shutdown.
i assume it must always happen in this sequence.
do we need to add any documentation to this method if these are done out of sequence

Reply:
Closer does everything synchronously and the first thing that happens is waiting for the latch to signal the fluxes have shut down.  At that point its just cleanup.